File: examples/mcp-search-notes/node/src/repository.ts
import { readFileSync } from 'node:fs'
import type { SearchNotesInput, SearchNotesOutput } from './contract.js'

interface StoredNote {
  id: string
  title: string
  body: string
}

export interface NoteRepository {
  search(input: SearchNotesInput): Promise<SearchNotesOutput>
}

export class FixtureNoteRepository implements NoteRepository {
  private readonly notes: StoredNote[]

  constructor() {
    const fixtureUrl = new URL('../../fixtures/notes.json', import.meta.url)
    this.notes = JSON.parse(readFileSync(fixtureUrl, 'utf8')) as StoredNote[]
  }

  async search({ query, limit }: SearchNotesInput): Promise<SearchNotesOutput> {
    const term = query.toLowerCase()
    return {
      items: this.notes
        .filter((note) => `${note.title} ${note.body}`.toLowerCase().includes(term))
        .slice(0, limit)
        .map((note) => ({ id: note.id, title: note.title, excerpt: note.body })),
    }
  }
}

File: examples/mcp-search-notes/python/src/search_notes/repository.py
import json
from pathlib import Path
from typing import Protocol

from .models import NoteResult, SearchNotesResult


class NoteRepository(Protocol):
    def search(self, query: str, limit: int) -> SearchNotesResult: ...


class FixtureNoteRepository:
    def __init__(self) -> None:
        fixture_path = Path(__file__).parents[3] / "fixtures" / "notes.json"
        self._notes: list[dict[str, str]] = json.loads(fixture_path.read_text())

    def search(self, query: str, limit: int) -> SearchNotesResult:
        term = query.casefold()
        visible = [
            NoteResult(id=note["id"], title=note["title"], excerpt=note["body"])
            for note in self._notes
            if term in f'{note["title"]} {note["body"]}'.casefold()
        ]
        return SearchNotesResult(items=visible[:limit])
